include/sampleflow/consumers/average_cosinus.h

```diff
@@ -0,0 +1,205 @@
+  * apart by specific index. In this sense, we get function where domain is bounded unsigned int and for
+  * each value we calculate cosinus average.
+  * Let's say unsigned int is equal to l and n_samples=n and n is bigger than l. Than formula can be written as:
+ * $\hat{cos(\theta_n)(l)}=\frac{1}{n-l}\sum_{t=1}^{n-l}{(\bm{x}_{t+l}^T\bar\bm{x})(||bm{x}_{t}||*||\bar\bm{x}||)}.
+ *
+ * This code for every new sample updates $\hat{cos(\theta_n)(l)}, l=1,2,3...,cosinus_length. Choice of
+ * cosinus length can done by setting it in mcmc_test.cc.
+ *
+ * Notice, that for each new sample $x_n$, we need to take a sample, that was l earlier than new sample ($x_{n-l}).
+ * Working this way, updating algorithm becomes very similar to mean_value.h algorithm.
+ *
+ * Everytime, for updation we calculate corresponding fractions. There numerator is dot product, while denominator -
+ * multiplied those two vector norms
+ *
+ * ### Threading model ###
+ *
+ * The implementation of this class is thread-safe, i.e., its
+ * consume() member function can be called concurrently and from multiple
+ * threads.
+   **/
+
+    template <typename InputType>
+    class AverageCosinus: public Consumer<InputType>
+    {
+      public:
+        /**
+         * The data type of the elements of the input type.
+         */
+
+       using scalar_type = typename InputType::value_type;
+
+       /**
+   	 * The data type, where we save multiple set of past samples.
+   	 */
+
+       using value_type = boost::numeric::ublas::matrix<scalar_type>;
+
+        /**
+         * Constructor.
+         */
+
+        AverageCosinus(unsigned int length);
+
+        /**
+         * Process one sample by updating the previously computed covariance
+         * matrix using this one sample.
+         *
+         * @param[in] sample The sample to process.
+         * @param[in] aux_data Auxiliary data about this sample. The current
+         *   class does not know what to do with any such data and consequently
+         *   simply ignores it.
+         */
+        virtual
+        void
+        consume (InputType     sample,
+                 AuxiliaryData aux_data) override;
+
+        /**
+         * A function that returns the average cosinus vector computed from the
+         * samples seen so far. If no samples have been processed so far, then
+         * a default-constructed object of type InputType will be returned.
+         *
+         * @return The computed average cosinus vector.
+         */
+        std::vector<scalar_type> get() const;
+
+      private:
+        /**
+         * A mutex used to lock access to all member variables when running
+         * on multiple threads.
+         */
+
+      mutable std::mutex mutex;
+
+
+        /**
+         * Parts for running average_cosinus
+         * Description of these parts is given above
+         */
+
+      std::vector<scalar_type> current_avg_cosinus;//Current average cosinus
```

Comment:
no need to document this twice: what you have after the `//` should simply be in the `/** ... */` block